types/set_test.go

```diff
@@ -0,0 +1,843 @@
+					if gotErr.Error() != test.expectedErr {
+						t.Errorf("Expected error to be %q, got %q", test.expectedErr, gotErr.Error())
+						return
+					}
+				}
+				t.Errorf("Unexpected error: %s", gotErr.Error())
+				return
+			}
+			if gotErr == nil && test.expectedErr != "" {
+				t.Errorf("Expected error to be %q, got nil", test.expectedErr)
+				return
+			}
+			if diff := cmp.Diff(got, test.expected); diff != "" {
+				t.Errorf("Unexpected diff (-expected, +got): %s", diff)
+			}
+		})
+	}
+}
+
+func TestSetTypeEqual(t *testing.T) {
+	t.Parallel()
+
+	type testCase struct {
+		receiver SetType
+		input    attr.Type
+		expected bool
+	}
+	tests := map[string]testCase{
+		"equal": {
+			receiver: SetType{ElemType: StringType},
+			input:    SetType{ElemType: StringType},
+			expected: true,
+		},
+		"diff": {
+			receiver: SetType{ElemType: StringType},
+			input:    SetType{ElemType: NumberType},
+			expected: false,
+		},
+		"wrongType": {
+			receiver: SetType{ElemType: StringType},
+			input:    NumberType,
+			expected: false,
+		},
+		"nil": {
+			receiver: SetType{ElemType: StringType},
+			input:    nil,
+			expected: false,
+		},
+		"nil-elem": {
+			receiver: SetType{},
+			input:    SetType{},
+			// SetTypes with nil ElemTypes are invalid, and
+			// aren't equal to anything
+			expected: false,
+		},
+	}
+	for name, test := range tests {
+		name, test := name, test
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+
+			got := test.receiver.Equal(test.input)
+			if test.expected != got {
+				t.Errorf("Expected %v, got %v", test.expected, got)
+			}
+		})
+	}
+}
+
+func TestSetElementsAs_stringSlice(t *testing.T) {
+	t.Parallel()
+
+	var stringSlice []string
+	expected := []string{"hello", "world"}
+
+	diags := (Set{
+		ElemType: StringType,
+		Elems: []attr.Value{
+			String{Value: "hello"},
+			String{Value: "world"},
+		}}).ElementsAs(context.Background(), &stringSlice, false)
+	if diags.HasError() {
+		t.Errorf("Unexpected error: %s", diags)
+	}
+	if diff := cmp.Diff(stringSlice, expected); diff != "" {
+		t.Errorf("Unexpected diff (-expected, +got): %s", diff)
+	}
+}
+
+func TestSetElementsAs_attributeValueSlice(t *testing.T) {
+	t.Parallel()
+
+	var stringSlice []String
+	expected := []String{
+		{Value: "hello"},
+		{Value: "world"},
+	}
+
+	diags := (Set{
+		ElemType: StringType,
+		Elems: []attr.Value{
+			String{Value: "hello"},
+			String{Value: "world"},
+		}}).ElementsAs(context.Background(), &stringSlice, false)
+	if diags.HasError() {
+		t.Errorf("Unexpected error: %s", diags)
+	}
+	if diff := cmp.Diff(stringSlice, expected); diff != "" {
+		t.Errorf("Unexpected diff (-expected, +got): %s", diff)
+	}
+}
+
+func TestSetTypeValidate(t *testing.T) {
+	t.Parallel()
+
+	testCases := map[string]struct {
+		in            tftypes.Value
+		expectedDiags diag.Diagnostics
+	}{
+		"null": {
+			in: tftypes.NewValue(
+				tftypes.Set{
+					ElementType: tftypes.String,
+				},
+				nil,
+			),
+		},
+		"null-element": {
```

Comment:
This seems like an unknown element, not a null one?

Reply:
Good catch, copypasta 🍝  -- fixed in next commit.